service/header/sync_ranges.go

```diff
@@ -0,0 +1,148 @@
+package header
+
+import "sync"
+
+// ranges keeps non-overlapping and non-adjacent header ranges which are used to cache headers (in ascending order).
+// This prevents unnecessary / duplicate network requests for additional headers during sync.
+type ranges struct {
+	ranges []*_range
+	lk     sync.Mutex // no need for RWMutex as there is only one reader
+}
+
+// Head returns the highest ExtendedHeader in all ranges if any.
+func (rs *ranges) Head() *ExtendedHeader {
+	rs.lk.Lock()
+	defer rs.lk.Unlock()
+
+	ln := len(rs.ranges)
+	if ln == 0 {
+		return nil
+	}
+
+	head := rs.ranges[ln-1]
+	return head.Head()
+}
+
+// Add appends the new ExtendedHeader to existing range or starts a new one.
+// It starts a new one if the new ExtendedHeader is not adjacent to any of existing ranges.
+func (rs *ranges) Add(h *ExtendedHeader) {
+	head := rs.Head()
+
+	// short-circuit if header is from the past
+	if head != nil && head.Height >= h.Height {
+		// TODO(@Wondertan): Technically, we can still apply the header:
+		//  * Headers here are verified, so we can trust them
+		//  * PubSub does not guarantee the ordering of msgs
+		//    * So there might be a case where ordering is broken
+		//    * Even considering the delay(block time) with which new headers are generated
+		//    * But rarely
+		//  Would be still nice to implement
+		log.Warnf("rcvd headers in wrong order")
+		return
+	}
+
+	rs.lk.Lock()
+	defer rs.lk.Unlock()
+
+	// if the new header is adjacent to head
+	if head != nil && h.Height == head.Height+1 {
+		// append it to the last known range
+		rs.ranges[len(rs.ranges)-1].Append(h)
+	} else {
+		// otherwise, start a new range
+		rs.ranges = append(rs.ranges, newRange(h))
+
+		// it is possible to miss a header or few from PubSub, due to quick disconnects or sleep
+		// once we start rcving them again we save those in new range
+		// so 'Syncer.getHeaders' can fetch what was missed
+	}
+}
+
+// BackWithin searches for a range within a given height span (start:end].
+func (rs *ranges) BackWithin(start, end uint64) (*_range, bool) {
+	r, ok := rs.Back()
+	if !ok {
+		return nil, false
+	}
+
+	if r.Start >= start && r.Start < end {
+		return r, true
+	}
+
+	return nil, false
+}
+
+// Back provides a first non-empty range, while cleaning up empty ones.
+func (rs *ranges) Back() (*_range, bool) {
```

Comment:
FirstNonEmptyRange  instead of Back?

Reply:
Too long. I would prefer to keep one word, max two.

Reply:
Can be also Tail, Dequeue, Last (no need to mention cleaning of empty ones in the name)

Reply:
I agree need other word. Prefer `FirstNonEmptyRange`

Reply:
Are you kidding me? It's not `First` by the code and by the other options I provided...

Reply:
`LastRange`? We don't return empty ranges for other methods as well, should I append `NonEmpty` to each? Obviously no, this and for this one, we don't need this.

Reply:
Oh, true, thanks for catching this. My bad that I confused you.

Reply:
Actually, from another look, it makes more sense to call `First`. Sorry for being defensive here and you were right about First

Reply:
Changed